# Анализ базы данных сервиса для чтения книг по подписке

Компания купила крупный сервис для чтения книг по подписке. 

Наша задача — проанализировать базу данных, содержащую информацию о книгах, издательствах, авторах, а также пользовательские обзоры книг. 


**Цель:**

Анализ базы данных для формулирования ценностного предложения для нового продукта.

**Описание данных**

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Создаем подключение к базе данных

In [1]:
# импортируем библиотеки
import pandas as pd 
from sqlalchemy import create_engine 

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net', 
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'], 
                                                         db_config['pwd'], 
                                                         db_config['host'], 
                                                         db_config['port'], 
                                                         db_config['db']
                                                        ) 

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

Импортировали все необходимые библиотеки и задали подключение. Переходим к изучению базы данных.

## Изучение базы данных

### таблица books

In [2]:
# составляем SQL запрос и сохраняем его в переменной
query = """
        SELECT *
        FROM books
        LIMIT 5;
        """
# выполняем SQL запрос
answer = pd.io.sql.read_sql(query, con = engine)
answer

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

### таблица authors

In [3]:
# составляем SQL запрос и сохраняем его в переменной
query = """
        SELECT *
        FROM authors
        LIMIT 5;
        """
# выполняем SQL запрос
answer = pd.io.sql.read_sql(query, con = engine)
answer

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

### таблица publishers

In [4]:
# составляем SQL запрос и сохраняем его в переменной
query = """
        SELECT *
        FROM publishers
        LIMIT 5;
        """
# выполняем SQL запрос
answer = pd.io.sql.read_sql(query, con = engine)
answer

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

### таблица ratings

In [5]:
# составляем SQL запрос и сохраняем его в переменной
query = """
        SELECT *
        FROM ratings
        LIMIT 5;
        """
# выполняем SQL запрос
answer = pd.io.sql.read_sql(query, con = engine)
answer

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

### таблица reviews

In [6]:
# составляем SQL запрос и сохраняем его в переменной
query = """
        SELECT *
        FROM reviews
        LIMIT 5;
        """
# выполняем SQL запрос
answer = pd.io.sql.read_sql(query, con = engine)
answer

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

Данные изучили. Они в поредке. Соответствуют описанию, которое мы получили. Переходим к анализу базы данных.

## Анализ базы данных

### считаем количество книг, вышедших после 1 января 2000 года:

In [7]:
# составляем SQL запрос и сохраняем его в переменной
query = """
        SELECT COUNT(book_id) as books_cnt
        FROM books
        WHERE publication_date > '2000-01-01';
        """
# выполняем SQL запрос
answer = pd.io.sql.read_sql(query, con = engine)
answer

books_cnt
0        819

После 1 января 2000 года вышло 819 книг.

### посчитаем количество обзоров и среднюю оценку для каждой книги:

In [8]:
# составляем SQL запрос и сохраняем его в переменной
query = """
        SELECT title,
               COUNT(DISTINCT review_id) AS review_cnt,
               ROUND(AVG(rating), 2) AS avg_rating
        FROM books AS b 
        LEFT JOIN ratings AS ra ON b.book_id = ra.book_id
        LEFT JOIN reviews AS re ON b.book_id = re.book_id
        GROUP BY b.book_id
        ORDER BY avg_rating DESC, 
                 review_cnt DESC;
        """
# выполняем SQL запрос
answer = pd.io.sql.read_sql(query, con = engine)
answer

title  review_cnt  avg_rating
0                        A Dirty Job (Grim Reaper  #1)           4        5.00
1              School's Out—Forever (Maximum Ride  #2)           3        5.00
2         Moneyball: The Art of Winning an Unfair Game           3        5.00
3                   The Big Bad Wolf  (Alex Cross  #9)           2        5.00
4                 Welcome to Temptation (Dempseys  #1)           2        5.00
..                                                 ...         ...         ...
995  The World Is Flat: A Brief History of the Twen...           3        2.25
996                                      Drowning Ruth           3        2.00
997                  His Excellency: George Washington           2        2.00
998                                              Junky           2        2.00
999                               Harvesting the Heart           2        1.50

[1000 rows x 3 columns]

Мы видим все книги, хранящиеся в базе данных, с их информацией о среднем рейтинге и количеству обзоров, отсортированные соответственно по рейтингу и количеству обзоров.


Книга с самым высокоим рейтингом и большим числом обзоров - это `A Dirty Job (Grim Reaper #1)`;

Самый низкий рейтинг и малое количество обзоров у книги `Harvesting the Heart`.	

### Определим издательство, которое выпустило наибольшее число книг толще 50 страниц (исключаем брошюры):

In [9]:
# составляем SQL запрос и сохраняем его в переменной
query = """  
        SELECT publisher,
               COUNT(book_id) AS books_cnt
        FROM publishers AS p
        INNER JOIN books AS b ON p.publisher_id = b.publisher_id
        WHERE num_pages > 50
        GROUP BY publisher
        ORDER BY books_cnt DESC
        LIMIT 1;                             
        """
# выполняем SQL запрос
answer = pd.io.sql.read_sql(query, con = engine)
answer

publisher  books_cnt
0  Penguin Books         42

Издательство `Penguin Books` выпустило больше всего книг, а именно 42.

### Определим автора с самой высокой средней оценкой книг (книги с 50 и более оценками):

In [10]:
# составляем SQL запрос и сохраняем его в переменной
query = """  
        SELECT author,
               ROUND(AVG(rating), 2) AS avg_rating
        FROM authors AS a
        JOIN books AS b ON a.author_id = b.author_id
        JOIN ratings AS r ON b.book_id = r.book_id
        WHERE b.book_id IN (
                            SELECT book_id
                            FROM ratings 
                            GROUP BY book_id
                            HAVING COUNT(rating_id) >= 50
                            ORDER BY COUNT(rating_id) DESC
                            )
        GROUP BY author
        ORDER BY avg_rating DESC
        LIMIT 1;                             
        """
# выполняем SQL запрос
answer = pd.io.sql.read_sql(query, con = engine)
answer

author  avg_rating
0  J.K. Rowling/Mary GrandPré        4.29

Видим, что автор с самой высокой средней оценкой книг (c 50 и более оценками) - это `J.K. Rowling/Mary GrandPré` со средней оценкой книг равной `4.29`.

### Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок:

In [11]:
# составляем SQL запрос и сохраняем его в переменной
query = """  
        SELECT ROUND(AVG(review_cnt), 2) as avg_review_cnt
        FROM (
              SELECT username, 
                     COUNT(review_id) AS review_cnt
              FROM reviews 
              WHERE username IN (
                                 SELECT username
                                 FROM ratings
                                 GROUP BY username
                                 HAVING COUNT(rating_id) > 50
                                 )
              GROUP BY  username
              ) AS counter;                             
        """
# выполняем SQL запрос
answer = pd.io.sql.read_sql(query, con = engine)
answer

avg_review_cnt
0           24.33

Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок равно `24.33`.

## Выводы

Мы провели анализ базы данных и выявили следующие моменты:

1. С 1 января 2000 года было выпущено 819 книг;
2. Книга с самым высокоим совокупным рейтингом и числом обзоров - это `A Dirty Job (Grim Reaper #1)`;
3. Больше всего книг выпустило издательство `Penguin Books` - 42;
4. Автор с самой высокой (4.29) средней оценкой книг (c 50 и более оценками) - это `J.K. Rowling/Mary GrandPré`;
5. Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок равно - `24.33`.